Bregman tv example

aringh opened the pull request:
A small Bregman-TV example in tomography.

However, I am not to convinced by the performance of the Bregman-TV algorithm on this example. Moreover, I would appreciate if anyone could double check that I have understood the idea behind iterative Bregman reconstruction correctly. Maybe @ozanoktem could have a look?

pep8speaks commented:
Checking new PR...

 - In the file [`examples/solvers/bregman_tv_tomography.py`](https://github.com/odlgroup/odl/blob/594760d1f182ff496ea640474f32bda3c4e9d78a/examples/solvers/bregman_tv_tomography.py), following are the PEP8 issues :

> [Line 29:18](https://github.com/odlgroup/odl/blob/594760d1f182ff496ea640474f32bda3c4e9d78a/examples/solvers/bregman_tv_tomography.py#L29): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator

 - Complete extra results for this file :

> noise = noise * 1/noise.norm() * noise_free_data.norm() * 0.10                 ^---

ozanoktem commented:
I have together with Martin Benning looked at the example and Martin sees some issues that can explain the poor results.
 
1. It seems that the sub gradient in the Bregman iterates is not updated, probably it stays to zero. Consequently, assuming constant initialisation as a sub-gradient, the first Bregman iterate is according to theory, but not the following ones. The correct updates are of the form as in Algorithm 1, page 32 in [https://arxiv.org/abs/1801.09922](https://arxiv.org/abs/1801.09922).

2. It seems easiest to have a functionality to update the sub-gradient at the discretion of the user. It is unclear from the provided example whether this is possible as indicated above.

adler-j commented:
Is there an open reference implementation that we could look at to find differences?

ozanoktem commented:
Martin Benning will provide clean Matlab code illustrating the algorithm, I will make sure it is available.

aringh commented:
Great! Thank you for the feedback.

Now, correct me if I am wrong, but this how I have understood it: to approximate a solution to the problem `min f(x) + J(x)`, a solution which has some certain nice features, we use Bregman iterations and stop after a few iterations.

For the Bregman iterations each iterate `x_{k+1}` is obtained by solving `min f(x) + D_J^{p_k}(x, x_k)`, where `D_J^{p_k}` is the Bregman distance of the functional `J` in the base point `x_k` and using the subgradient `p_k`. As I have understood one want the subgradient `p_{k}` to be a subgradient of `J` in the base point `x_k`, correct?

In an iterative scheme I can compute the new subgradient `p_{k+1}` using the previous subgradient, as given in the algorithm on page 32. But if I have a way to evaluate the subgradient of the functional `J`, am I still forced to use this update rule?

What the current code does is simply to in each step of the Bregman iterations initialize a new Bregman functional at the new base point `x_k` (the code line `bregman_func = reg_param * l1_norm.bregman(gradient(x))`). This initialization of course includes computing a subgradient of the functional in the given base point. Since the subgradient operator is not necessarily single valued, this may of course differ from the one in the algorithm in page 32. But is there a reason why the algorithm performance should deteriorate in this case?

Or, have I misunderstood how the Bregman iterations work?

ozanoktem commented:
Attached MATLAB code is from Martin Benning and it is used to create the deconvolution Bregman example in Section 6.1 in [https://arxiv.org/abs/1801.09922](https://arxiv.org/abs/1801.09922). You only need to unzip the content of the attached file into a folder, add the sub-folders to the Matlab path (either manually or preferably with the file setpath.m) and run the Matlab live script deconvolution.mlx in the folder `Examples/Deconvolution.` This file also contains step-by-step comments.

[BregmanTV_example.zip](https://github.com/odlgroup/odl/files/1800120/BregmanTV_example.zip)

ozanoktem commented:
Dear @aringh ,

You are correct in assuming that you can select sub-gradients different from how its been done in, for example, [https://arxiv.org/abs/1801.09922](url). If the underlying functional is differentiable, you do not even need to select one at all, as the sub-differential is single-valued. However, many theoretical guarantees (such as Lemma 6.1 in [https://arxiv.org/abs/1801.09922](url)) rely on the specific sub-gradient-descent-type update. This could therefore explain why your results appear to be different compared to Bregman iterations with the conventional update. 
In addition to that, I'm wondering how you pick a specific sub-gradient. Do you use something like p = -div sign( nabla u), where sign is a vector-valued sign function that returns zero for each pixel where nabla u(x) = 0? 

Cheers,

Martin

mbenning commented:
Oops, I accidentally was still signed in as Ozan. Responses and complaints about the previous post have to be redirected to me ;-)

adler-j commented:
Is there any chance that we could put some effort into finishing this? I'd love to have a proper bregman example inside ODL.

mehrhardt commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,91 @@
+# Create sinogram of forward projected phantom with noise
+noise_free_data = ray_trafo(discr_phantom)
+noise = odl.phantom.white_noise(ray_trafo.range)
+noise = noise * 1/noise.norm() * noise_free_data.norm() * 0.10
```
small thing, but why not `noise *= (1/noise.norm() * noise_free_data.norm() * 0.10)`?

mehrhardt commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,91 @@
+geometry = odl.tomo.Parallel2dGeometry(angle_partition, detector_partition)
+
+# Create the ray transform
+ray_trafo = odl.tomo.RayTransform(reco_space, geometry, impl='astra_cuda')
```
I suppose these examples should not have CUDA hardcoded in there?

mehrhardt commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,91 @@
+
+# Discrete reconstruction space
+reco_space = odl.uniform_discr(min_pt=[-20, -20], max_pt=[20, 20],
+                               shape=[300, 300], dtype='float32')
```
to me this is way too "large" for a simple example.

mehrhardt commented:
Looks OK to me how you did it. Could be more elegant but there is no simple obvious choice how to do it. Why did you not implement the subgradient update as you quoted in your post?

aringh commented:
Thank you for the feedback! The reason I have not done it as I described above is because I am do not know how to do it :wink: 

The update of the subgradient as above will give me a subgradient that lives in the same space as the image (technically the dual space, but in this discretized setting it will effectively be something with the same discretization as the image). However, as the code is currently written, my Bregman functional is defined only for the L1-part of the regularize, and this means that the domain of this functional is the product space of the image space with itself. Therefore, the subgradient I need for the Bregman functional needs to live in this space, and I have not been able to figure out how to go from a subgradient `p`, which is a subgradient of the composition of the L1-norm and the spatial gradient by its construction, to a subgradient for only the L1-norm.

The reason for constructing the Bregman functional only for the L1-part is that I want to do operator splitting when I solve the nonsmooth problem.

I can relatively easy fix this example to a working Bregman-TV example, but that is _without_ using the `.bregman`-functionality. And that is not really what I would like, since if I cannot do an easy example with Bregman-TV using this functionality, why do we have it?

aringh commented:
Well, that's good! :smiley: Then maybe I can just change this to not use the `.bregman` -functionality. In that case it should not be to hard.

aringh commented:
So, I have:

1. squashed the three previous commits, since the two last ones were just minor adjustments.
2. changed the example to not use the `.bregman`-functionality, but instead I manually keep track of the linear part. After some tuning, I think that this is now an ok example.
3. rebased on master.

The example is now up for review.

aringh commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,95 @@
+g = [l2_norm, reg_param*l1_norm]
+L = [ray_trafo, gradient]
+
+# Estimated operator norms, add 10 percent to ensure that we fulfill the
```
No longer add 10%, update comment according to code.

aringh commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,95 @@
+"""Example of Bregman-TV in tomography.
+
+The example does seven iterations in a Bregman-TV scheme, applied to a
```
Potentially add a longer description of the example? At least add a reference.

aringh commented:
This one is up for review if anyone is interested and have time :smiley:

adler-j commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,95 @@
+noise *= 1 / noise.norm() * noise_free_data.norm() * 0.10
+data = noise_free_data + noise
+
+# Do a FBP-reconstruction
```
Comments like these are quite un-specific. It's obvious from the code that we're doing a FBP reconstruiction, so either scrap the comment or add more value, e.g. "Do a FBP-reconsturction to great initial guess", or something similar.

kohr-h replied:
an

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,95 @@
+l2_norm = 0.5 * odl.solvers.L2NormSquared(ray_trafo.range).translated(data)
+
+# Isotropic TV-regularization
+l1_norm = odl.solvers.GroupL1Norm(gradient.range)
```
Maybe name this `l12_norm` to avoid confusion. Probably other examples have the same issue.

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,95 @@
+
+# Assembel functionals and operators for the optimization solver
+f = odl.solvers.ZeroFunctional(reco_space)  # Don't use f functional, set zero
+g = [l2_norm, reg_param*l1_norm]
```
Spaces around `*`

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,95 @@
+
+# Estimated operator norms, add 10 percent to ensure that we fulfill the
+# convergence criteria of the optomization algorithm
+ray_trafo_norm = odl.power_method_opnorm(ray_trafo)
```
You may want to add `maxiter` here for CPU users.

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,95 @@
+niter = 200  # Number of iterations in the inner loop
+tau = 0.01  # Step size for the primal variable
+sigma_ray_trafo = 45.0 / ray_trafo_norm**2  # Step size for the dual variable
+sigma_gradient = 45.0 / gradient_norm**2  # Step size for the dual variable
```
Spaces around `**`

kohr-h replied:
Also the numbers are a bit "magical". Why not use the default stepsize helper? (It may not be implemented for forward-backward, though...)
Adding a parameter check would be nice for users.

aringh replied:
Since it is an iterative procedure inside an iterative procedure, I wanted to not have to  do to many inner iterations. Therefore I have tuned the numbers in order to be able to reduce `niter` to "only" 200. So the numbers are indeed a bit magical.

Taking something more standard as standard `tau = 0.5` and `sigma_i = 1 / op_norm**2` seems to demand many more inner iterations.

kohr-h replied:
Ok, fair enough. That would definitely be worth mentioning in a comment.

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,95 @@
+gradient_norm = odl.power_method_opnorm(gradient)
+
+# Set parameters for the optimization algorithm
+niter = 200  # Number of iterations in the inner loop
```
A more obvious name would be `niter_inner`

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,95 @@
+p = reco_space.zero()
+
+# This defines the Bregman iterations
+for breg_iter in range(7):
```
Even though it's only in one place, making a variable `niter_bregman` makes it more clear / less magical.

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,95 @@
+        callback=callback_inner)
+
+    # Update the subgradient
+    p -= 1/reg_param * ray_trafo.adjoint(l2_norm.gradient(ray_trafo(x)))
```
Spaces around `/`

adler-j replied:
Put a parenthesis here to make it more clear, e.g. `(1/reg_param)`

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,95 @@
+import odl
+import numpy as np
+
+# Discrete reconstruction space
```
~Discrete~

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,95 @@
+# Display phantom, FBP reconstruction, and sinograms
+discr_phantom.show(title='Phantom')
+fbp_reco.show(title='FBP reconstruction')
+data.show(title='Simulated data (Sinogram)')
```
Maybe move these to the very end, together with the final Bregman result.

aringh replied:
I will move these to the end. But I will not make a final show of the Bregman recosntruction. Because the example is tuned so that something like 5 iterations is enough, but I do 7. This is on purpose, to illustrate that one needs to do early stopping when doing Bregman iterations. But maybe I should document that somewhere too.

adler-j commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,95 @@
+tau = 0.01  # Step size for the primal variable
+sigma_ray_trafo = 45.0 / ray_trafo_norm**2  # Step size for the dual variable
+sigma_gradient = 45.0 / gradient_norm**2  # Step size for the dual variable
+sigma = [sigma_ray_trafo, sigma_gradient]
```
Have you tried using the default choices?

aringh replied:
Yes, see answer to @kohr-h's comment on the same line

adler-j commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,95 @@
+# Create sinogram of forward projected phantom with noise
+noise_free_data = ray_trafo(discr_phantom)
+noise = odl.phantom.white_noise(ray_trafo.range)
+noise *= 1 / noise.norm() * noise_free_data.norm() * 0.10
```
Move 0.10 to the left

adler-j commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,95 @@
+# Make a parallel beam geometry with flat detector
+angle_partition = odl.uniform_partition(0, np.pi, 100)
+detector_partition = odl.uniform_partition(-30, 30, 558)
+geometry = odl.tomo.Parallel2dGeometry(angle_partition, detector_partition)
```
Simply go for `parallel_beam_geometry`

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,134 @@
+    p_{k+1} - p_k in -(1/lam) * A^* dD(Ax_k, g).
+
+Note that the first step involves solving an optimization problem, which is
+(normally) done using in iterative method. This means that the solution method
```
~in~ an

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,134 @@
+
+The example solves an inverse problem in tomography using Bregman iterations
+for a total variation (TV) regularized problem. To this end, let ``A`` denote
+the forward operator, ``g`` denote the given data, ``D(Ax, g)`` denote data
```
denote the data

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,134 @@
+                               shape=[128, 128], dtype='float32')
+
+# Make a parallel beam geometry with flat detector
+geometry = odl.tomo.parallel_beam_geometry(reco_space)
```
How many angles does this use? Could be quite overkill. In the interest of CPU users, consider setting number of angles to a fixed value, 100-ish. That would also be more interesting as an inverse problem.

aringh replied:
It seems to use 285 angles 183 detector bins. I updated since @adler-j commented and told me to simplify to this :stuck_out_tongue_closed_eyes: Before it was 
```
angle_partition = odl.uniform_partition(0, np.pi, 100)
detector_partition = odl.uniform_partition(-30, 30, 558)
geometry = odl.tomo.Parallel2dGeometry(angle_partition, detector_partition)
```
This update is also why I in this commit updated the regularization parameter; the change in geometry gave a less interesting result with the previous parameter.

kohr-h replied:
Oh but you can give `num_angles` to the function, no need to go back to the old one.

aringh replied:
Ah! I did not know :-)

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,134 @@
+ray_trafo = odl.tomo.RayTransform(reco_space, geometry)
+
+# Create phantom
+discr_phantom = odl.phantom.shepp_logan(reco_space, modified=True)
```
Can you try to make larger blocks of code with comments on top? This feels like lots of comments that state the obvious. For instance, the phantom could easily be part of the "create data" block, and the reco space, geometry and ray trafo could be one block as well.

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,134 @@
+# Create phantom
+discr_phantom = odl.phantom.shepp_logan(reco_space, modified=True)
+
+# Create sinogram of forward projected phantom. Add 10% Guassian noise
```
Dot or no dot? I tend to do it like this for comments, i.e., no trailing dot, but it gets harder with 2 sentences. In those cases I resort to semicolon instead of full stop, i.e., "Do this thing; do a second thing"

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,134 @@
+# Isotropic TV-regularization
+l12_norm = odl.solvers.GroupL1Norm(gradient.range)
+
+# Assembel functionals and operators for the optimization solver
```
Typo "Assembel"
And "optimization solver" is a tautology, either just "solver", or "optimization method", or "optimizer"

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,134 @@
+noise *= 0.10 / noise.norm() * noise_free_data.norm()
+data = noise_free_data + noise
+
+# Define components of the variational problem
```
This comment is for the whole section, but it looks like it only applies to `reg_param` and `gradient`. Another good reason to make larger blocks. If you need an intermediate comment in a block, I'd recommend doing it without blank line before, or, if it fits, on the same line:
```
# This is the summary of the block
some code running here
more code
some non-obvious code  # explained very briefly here
# This is a more elaborate comment on the next line
here's another non-obvious statement

# Next block starts here
```

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,134 @@
+l12_norm = odl.solvers.GroupL1Norm(gradient.range)
+
+# Assembel functionals and operators for the optimization solver
+f = odl.solvers.ZeroFunctional(reco_space)  # Don't use f functional, set zero
```
set to zero

aringh replied:
On purpose: I have only one character left on this line before I hit the 79 character limit :wink:. I will try to reformulate

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,134 @@
+niter_inner = 200  # Number of iterations in the inner loop
+tau = 0.01  # Step size for the primal variable
+sigma_ray_trafo = 45.0 / ray_trafo_norm**2  # Step size for the dual variable
+sigma_gradient = 45.0 / gradient_norm**2  # Step size for the dual variable
```
Spaces around `**` still missing

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,134 @@
+sigma_gradient = 45.0 / gradient_norm**2  # Step size for the dual variable
+sigma = [sigma_ray_trafo, sigma_gradient]
+
+# Number of Bregman iterations
```
Obvious comment. Better explain why 7.

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,134 @@
+# Number of Bregman iterations
+niter_bregman = 7
+
+# Create initial guess for the solver.
```
No dot

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,134 @@
+x = reco_space.zero()
+
+# Create initial subgradient
+p = reco_space.zero()
```
Do both `x` and `p` in one block

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,134 @@
+    linear_part = reg_param * odl.solvers.QuadraticForm(vector=-p,
+                                                        constant=constant)
+
+    # Used to print iteration number.
```
Obvious

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,134 @@
+    # Used to print iteration number.
+    callback_inner = odl.solvers.CallbackPrintIteration(step=50)
+
+    # Call the solver. x is updated in-place with the consecutive iterates.
```
Again the issue with full stops, would be nice to have consistency

kohr-h commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,134 @@
+# Set parameters for the optimization algorithm. These are tuned in order to
+# reduce the number of inner iterations needed to solve the first step in the
+# Bregman iteration.
+niter_inner = 200  # Number of iterations in the inner loop
```
Obvious comment

aringh commented:
I will merge after the tests have run

kohr-h commented:
> I will merge after the tests have run

Can you squash the commits? We don't really have an `EX` label :wink:

aringh commented:
I know we don't, but I created one since I thought that "ENH" was a bit to ambitious for adding an example. But I can change :wink:

kohr-h commented:
> I know we don't, but I created one since I thought that "ENH" was a bit to ambitious for adding an example. But I can change wink

I think adding an example is a clear ENH. The other one, "update..." makes it a bit hard to find a good label because it's not very specific.
But please don't change it again :grimacing:

aringh commented:
I wanted to keep the two commits separate explicitly, since the first one is a try to do it using `.bregman`. It never really worked, but in the spirit of keeping some of the development history I though it could be good to keep it if someone wants to try it again. (The probability is very low that someone will find the commit, but anyway.)

mehrhardt commented on examples/solvers/bregman_tv_tomography.py:
```diff
@@ -0,0 +1,125 @@
+    callback_inner = odl.solvers.CallbackPrintIteration(step=50)
+
+    # Inner iterations; x is updated in-place with the consecutive iterates
+    odl.solvers.forward_backward_pd(
```
Without the restrictions of ODL, I would relate `h` with the data fit. The update of the subgradient includes the gradient of the data fit at the last iterate. Thus, one could recycle these computations. It would be good if ODL had the option to do this. In particular for real size problems. This strategy would also be crucial for the linearized Bregman iterations.

kohr-h replied:
That will all be possible when #1400 lands.

aringh replied:
As of now, one can wrap both the data matching and the linear part of the Bregman functional in a `FunctionalQuadraticPerturb`, and pass all of that as the smooth part of the solver (i.e., as `h`). I guess it is something like that you were looking for?

However, before we have something like #1400 in place it will be hard (impossible?) to reuse the gradients computed by the solver.
